natural-language-classifier/v1-generated.ts

```diff
@@ -0,0 +1,296 @@
+      })
+    };
+    return requestFactory(parameters, callback);
+  };
+
+  /**
+   * Delete classifier.
+   *
+   *
+   * @param {Object} params - The parameters to send to the service.
+   * @param {string} params.classifier_id - Classifier ID to delete.
+   * @param {Function} [callback] - The callback that handles the response.
+   */
+  deleteClassifier(params: GeneratedNaturalLanguageClassifierV1.DeleteClassifierParams, callback?: GeneratedNaturalLanguageClassifierV1.Callback<GeneratedNaturalLanguageClassifierV1.Empty>): ReadableStream | void {
+    const requiredParams = ['classifier_id'];
+    const missingParams = helper.getMissingParams(params || {}, requiredParams);
+    if (missingParams && callback) return callback(missingParams);
+    const path = { classifier_id: params.classifier_id };
+    const parameters = {
+      options: {
+        url: '/v1/classifiers/{classifier_id}',
+        method: 'DELETE',
+        path: path,
+      },
+      defaultOptions: extend(true, this._options, {
+        headers: {
+          'accept': 'application/json',
+          'content-type': 'application/json',
+        }
+      })
+    };
+    return requestFactory(parameters, callback);
+  };
+
+  /**
+   * Get information about a classifier.
+   *
+   * Returns status and other information about a classifier.
+   *
+   * @param {Object} params - The parameters to send to the service.
+   * @param {string} params.classifier_id - Classifier ID to query.
+   * @param {Function} [callback] - The callback that handles the response.
+   */
+  getClassifier(params: GeneratedNaturalLanguageClassifierV1.GetClassifierParams, callback?: GeneratedNaturalLanguageClassifierV1.Callback<GeneratedNaturalLanguageClassifierV1.Classifier>): ReadableStream | void {
+    const requiredParams = ['classifier_id'];
+    const missingParams = helper.getMissingParams(params || {}, requiredParams);
+    if (missingParams && callback) return callback(missingParams);
+    const path = { classifier_id: params.classifier_id };
+    const parameters = {
+      options: {
+        url: '/v1/classifiers/{classifier_id}',
+        method: 'GET',
+        path: path,
+      },
+      defaultOptions: extend(true, this._options, {
+        headers: {
+          'accept': 'application/json',
+          'content-type': 'application/json',
+        }
+      })
+    };
+    return requestFactory(parameters, callback);
+  };
+
+  /**
+   * List classifiers.
+   *
+   * Returns an empty array if no classifiers are available.
+   *
+   * @param {Object} [params] - The parameters to send to the service.
+   * @param {Function} [callback] - The callback that handles the response.
+   */
+  listClassifiers(params?: GeneratedNaturalLanguageClassifierV1.ListClassifiersParams, callback?: GeneratedNaturalLanguageClassifierV1.Callback<GeneratedNaturalLanguageClassifierV1.ClassifierList>): ReadableStream | void {
+    params = params || {};
+    if (typeof params === 'function' && !callback) {
```

Comment:
🤔

Reply:
I think we always want to have parameters and callback.

Reply:
Remember that `list()` could have parameters in the future

Reply:
This is definitely awkward. I really wish JavaScript supported function overloading.

It's a fair point that new parameters may be added to the `listClassifiers` function. A new _required_ parameter will necessitate a breaking change for all SDKs, regardless. However, a new _optional_ parameter would still keep the entire `params` object as optional.

Personally, I'd rather embrace the _internal_ awkwardness of checking for the `params` argument rather than the _external_ awkwardness of requiring users to pass a `params` argument of `{}` or `undefined`.

Would be helpful to get @nfriedly's opinion when he's back from his trip.

Reply:
For consistency, we should stick with `(params, callback)` for all our methods.
You can leave it if you want but it won't be documented.
In Java, we had a similar situation, and we decided to send a model in the request even if it was empty or null.

Reply:
I talked to Mike to learn more about this situation in Java. While chatting, we realized that we could create an overloaded method in Java:

```java
// original method -- options has no required parameters
public ServiceCall<Model> listModels(Options options) {
    ...
}

// new overloaded method
public ServiceCall<Model> listModels() {
    listModels(null);
}
```

This was a quick and easy change to the Java service template. Mike is going to test it out and if all goes well then he'll open a pull request in `watson-swagger-codegen`. Any generated code should be backwards-compatible with the existing Java SDK release since it doesn't modify the existing method.

In light of that change, I think it makes sense to try the equivalent of an overload in Node. I think the closest we can get is to check the arguments and swizzle them around, as awkward as it is.

Reply:
Which method do we document the overloaded or the other one?

Reply:
I think they both have to be documented.

For example, consider a `listModels` endpoint that only has _optional_ parameters, so both methods will be generated. The `listModels()` function with no arguments may be the common path, so it should be documented. But users that want to pass specific options may choose to use `listModels(ListModelsOptions options)`. So it should be documented as well.

Reply:
I like the code as it's written - with optional params.  

For the documentation, I think that the typescript typings and JSDoc should accurately reflect that it is optional, but any hand-written examples should probably pass in an options object for consistency.

The API ref doesn't explicitly mention the options object for Node.js, just it's keys, so it woon't really make any difference there aside from the code example.

Reply:
@nfriedly How do you expect us to support promises without breaking users? People in the community don't like to support callbacks and promises in the same function.

Reply:
Regarding the optional parameter, I didn't hear a strong argument of why we should do it so let's just not do it. Especially since we are going to document examples where we are going to pass empty objects.

Reply:
Good discussion, but let's leave it optional. That seems more idiomatic for Node. And for what it's worth, this is the Node equivalent of method overloading in the [Java template][1], so it encourages consistency between languages.

[1]: https://github.ibm.com/MIL/watson-swagger-codegen/blob/develop/src/main/resources/watson-java/service.mustache#L360

Reply:
I agree with you. Let's use your approach and make sure we don't document it in the API Reference for now